Homework 4 - Patterns and Refactoring

fdemchenko opened the pull request:
1. Add one more exchange rate provider
2. Create cache with expiration
3. Rewrite rate service using cache and different providers
4. Repository pattern was already implemented
5. Add logging to rate service

fdemchenko committed:
Remove revive and ireturn linters

fdemchenko committed:
Implement cache with expiration

fdemchenko committed:
Refactor rate service

Change the way rate service caches fetched rate, know it uses my own
cache implementation, instead of backgound task.
Add one more currency exchange rate provider, so now service will
fallback to it in case first one is not responding.
Change rate service unit tests.

fdemchenko committed:
Add tests block to README.md

fdemchenko committed:
Fix NBU unmarshaling, add logging to rate service

fdemchenko committed:
Fix errors

MaxDMT16 commented on internal/services/rate.go:
This files looks a bit overloaded with different functionalities:
- NBU fetching logic
- Fawaz fetching logic
- rate service with caching and it's options

I think split these code into separate files would improve readability.
e.g.

- /services
  - /rate
    - rate.go - rate service with caching and it's options
    - fawaz.go - Fawaz fetching logic
    - nbu.go - NBU fetching logic

DimaL-cloud replied:
+, rate service violates OCP

fdemchenko replied:
Absolutely agree, will split rate specific code into separate package

MaxDMT16 commented on internal/services/rate.go:
```diff
@@ -1,115 +1,160 @@
-type Rate struct {
+var ErrInvalidCurrencyCode = errors.New("invalid currency code")
+
+type ExchangeRateFetcher func(code string, client *http.Client) (float32, error)
```
There is a convention to name delegates like this `SomeActionFunc`
e.g. `FetchExchangeRateFunc`

MaxDMT16 commented on internal/services/rate.go:
```diff
@@ -1,115 +1,160 @@
+}
+
+func NBURateFetcher(code string, client *http.Client) (float32, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), RateFetchTimeout)
```
What's the benefit of using `context.Background()` in such cases (in fawaz fetcher the same approach)? Why not to use request's context?

fdemchenko replied:
I'm using `context.Background()` as placeholder, to add timeouts to requests. Indeed, passing request's context down the calls chain is the right way, and is the reason context package was created. I'll handle it.

MaxDMT16 commented on internal/services/rate.go:
```diff
@@ -1,115 +1,160 @@
 }
 
-var ErrNoFetchOccurred = errors.New("no fecth occurred yet")
+func FawazAhmedRateFetcher(code string, client *http.Client) (float32, error) {
```
It's cool that you've extended rate fetcher with currency code.
However you could make it even better by parameterizing both `from` and `to` currencies

fdemchenko replied:
For now, not all providers provide info about all the currencies, but I'll work on it!

MaxDMT16 commented on internal/services/rate.go:
```diff
@@ -1,115 +1,160 @@
 
-type httpExchangeRateClient struct {
-	client *http.Client
+func CreateNamedFetcher(fetcher ExchangeRateFetcher, name string) NamedFetcher {
+	return NamedFetcher{fetcher: fetcher, name: name}
 }
```
It's seems illogical to me that you should specify fetcher name each time as it remains the same for fetcher of the same time across the project. I suggest to incapsulate name setting in fetcher and just expose `Name()` func that will return it to use outside

MaxDMT16 commented on internal/services/rate.go:
```diff
@@ -1,115 +1,160 @@
+	var err error
+	var rate float32
+	for _, namedFetcher := range crs.namedFetchers {
+		rate, err = namedFetcher.fetcher(currencyCode, crs.client)
```
`namedFetcher.fetch(...)`

MaxDMT16 commented on internal/services/rate_test.go:
```diff
@@ -13,46 +14,70 @@ const (
-	_, _ = rateService.GetRate()
-	_, _ = rateService.GetRate()
+	mockRateFetcher := new(MockExchangeRateFetcher)
+	mockRateFetcher.On("FirstFetcher").Return(float32(8.0), nil)
```
Please declare constants for literals

MaxDMT16 commented on internal/services/rate_test.go:
```diff
@@ -13,46 +14,70 @@ const (
+	mockRateFetcher.AssertNumberOfCalls(t, "FirstFetcher", 2)
+}
+
+func TestRateService_FallbackToAnotherFetcher(t *testing.T) {
```
Unfortunately, this test doesn't pass successfully(
![image](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-fdemchenko/assets/24816241/6e777e8d-4954-407c-bca5-dc3628cfcc72)


I suppose the root cause is the way [`WithFetchers`](https://github.com/GenesisEducationKyiv/software-engineering-school-4-0-fdemchenko/pull/5/files#diff-5ed4d654fee99e78ea02f2a6c861077af36b53073891be54b973af158d060c8bR122-R125) option is implemented. Could you please fix it?

fdemchenko replied:
Yeah, sure. Fixed!

MaxDMT16 reviewed (changes requested):
In general good!
It's cool that you've added options pattern 👍🏼

However, there are few points to fix:
- It's said in HA task to add several currency providers, but you've added just one.
Please try to add one more. It will give ability to test the approach you implemented in `rateService`.
- one of the tests fails. It's a bad practise to merge changes when any test is red) Please fix it

fdemchenko committed:
Refactor rate service

Now fetcher is a struct not a function, with name property.
Split rate specific code into separate package. Add new currency
provider (PrivatBank API). Allow client's code to pass context to
GetRate() method.

fdemchenko committed:
Update rate service unit tests

fdemchenko committed:
Update server integration test

fdemchenko committed:
Extract literals into constant

fdemchenko committed:
Change types in Privat response, add float32 parsing. Fix minor linter issues

MaxDMT16 commented on internal/services/rate/privat.go:
```diff
@@ -0,0 +1,62 @@
+)
+
+const (
+	PrivatExchangeRateURL = "https://api.privatbank.ua/p24api/pubinfo?json&exchange&coursid=5"
```
Are you inspired by Dmytro's PR?)

fdemchenko replied:
What kind of inspiration?)

MaxDMT16 replied:
You've used the same providers as Dmytro)

fdemchenko replied:
Just because these providers are completely free. But unfortunately, not all of them provide full currency list data

MaxDMT16 reviewed (approved):
Well done 🙌🏼
Looks much better!
